### ISSUE TITLE
`accept()` code-path doesn't set socket to `O_NONBLOCK`, resulting in blocking reads

### ISSUE THREAD

**Issue by jshearer** [tokio-rs/mio#1450]
I spent a good couple of days debugging this one, and I learned so much! It appears that a while ago, [a commit](https://github.com/tokio-rs/mio/commit/b775034106ac402596ad72cedf376f613edc6268) was introduced that assumes that the file descriptors provided by `accept()` will inherit their parent socket's blocking attribute. This... seems to not be the case on Android, as after collecting a few `strace` outputs, I discovered that `recvfrom()` calls were blocking where they very much shouldn't have been, resulting in all sorts of trouble. 

```
##### SOCKET 93 CREATED HERE
[pid  6747] socket(AF_UNIX, SOCK_STREAM, 0 <unfinished ...>
[pid  6747] <... socket resumed> )      = 93

#### SET TO NONBLOCK HERE
[pid  6747] fcntl64(93, F_SETFL, O_RDONLY|O_NONBLOCK <unfinished ...>
[pid  6747] <... fcntl64 resumed> )     = 0
[pid  6747] fcntl64(93, F_SETFD, FD_CLOEXEC <unfinished ...>

###### BINDS HERE 
[pid  6747] bind(93, {sa_family=AF_UNIX, sun_path="xxx"}, 108 <unfinished ...>
[pid  6747] <... bind resumed> )        = 0

###### LISTENS HERE 
[pid  6747] listen(93, 1024 <unfinished ...>
[pid  6747] <... listen resumed> )      = 0

###### CONNECTION ACCEPTED HERE, SWITCH OVER TO FD 101
[pid  6746] accept4(93,  <unfinished ...>
[pid  6746] <... accept4 resumed> {sa_family=AF_UNIX}, [110->2], 0) = 101
[pid  6746] fcntl64(101, F_SETFD, FD_CLOEXEC <unfinished ...>
[pid  6746] <... fcntl64 resumed> )     = 0

##### DIES HERE
[pid  6746] recvfrom(101,  <unfinished ...>

#### NEVER RESUMES
```

After testing a patch locally that sets the resulting fd's `O_NONBLOCK` flag, this blocking turned nonblocking as expected. All is right in the world!

Reading the [docs](https://man7.org/linux/man-pages/man2/accept.2.html), it appears that 
> On Linux, the new socket returned by accept() does _not_ inherit
       file status flags such as `O_NONBLOCK` and `O_ASYNC` from the
       listening socket.  This behavior differs from the canonical BSD
       sockets implementation.  Portable programs should not rely on
       inheritance or noninheritance of file status flags and always
       explicitly set all required flags on the socket returned from
       `accept()`.

Even the NetBSD [docs](https://man.netbsd.org/accept.2#COMPATIBILITY) say
>  The `accept()` implementation makes the new file descriptor inherit file
     flags (like O_NONBLOCK) from the listening socket.  It's a traditional
     behaviour for BSD derivative systems.  On the other hand, there are
     implementations which don't do so.  Linux is an example of such implemen-
     tations.  **Portable programs should not rely on either of the behaviours.**

Emphasis mine

**Comment by Thomasdezeeuw** [comment]
On Android (not x86) we call `accept4(2)` with `SOCK_NONBLOCK`, see: https://github.com/tokio-rs/mio/blob/527e843b0e11d585c0b20a8a3e4df966766f8bfa/src/sys/unix/uds/listener.rs#L41-L62

That should set the accepted stream to non-blocking, but apparently on Android it doesn't. I think we'll need some more information about the Android setup. Also could you run the test suite of Mio on the Android device, without your changes?

**Comment by jshearer** [comment]
Hi @Thomasdezeeuw! Sorry, I suppose I should have clarified: this issue happens on x86 Android. Attached is the test output, where the 3 tests hang indefinitely. Are you saying that since it's _just_ Android that's affected, it doesn't make sense to make these syscalls on the other platforms that are in the codepath (ios, macos, netbsd, solaris)? If so I can certainly gate it behind `#[cfg(and(target_arch="x86", target_os="android"))]`, I just figured that since both the docs I found for this syscall recommend always being explicit in setting the options, it would be the right thing to do. Thoughts?

Android x86 test output (I manually killed the tests after a few minutes, they were clearly blocking):

```
running 4 tests
test poll::as_raw_fd ... ok
test sys::unix::selector::epoll::assert_close_on_exec_flag ... ok
test sys::unix::uds::tests::abstract_address ... ok
test sys::unix::uds::tests::pathname_address ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/aio-bdcb55700b53709f`
/Users/js/mio/target/i686-linux-android/debug/deps/aio-bdcb55700b53709f: 1 file pushed, 0 skipped. 40.1 MB/s (4935836 bytes in 0.117s)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/close_on_drop-9ad1c01f04edbb4e`
/Users/js/mio/target/i686-linux-android/debug/deps/close_on_drop-9ad1c01f04edbb4e: 1 file pushed, 0 skipped. 31.3 MB/s (6712656 bytes in 0.204s)

running 1 test
test close_on_drop ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/events-33c4dc14eaec30ad`
/Users/js/mio/target/i686-linux-android/debug/deps/events-33c4dc14eaec30ad: 1 file pushed, 0 skipped. 36.4 MB/s (6628780 bytes in 0.174s)

running 2 tests
test assert_event_source_implemented_for ... ok
test events_all ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/interest-87e366adc6ddb09e`
/Users/js/mio/target/i686-linux-android/debug/deps/interest-87e366adc6ddb09e: 1 file pushed, 0 skipped. 28.6 MB/s (5724472 bytes in 0.191s)

running 4 tests
test add ... ok
test bit_or ... ok
test fmt_debug ... ok
test is_tests ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/poll-451ef23ffc224ef0`
/Users/js/mio/target/i686-linux-android/debug/deps/poll-451ef23ffc224ef0: 1 file pushed, 0 skipped. 27.2 MB/s (7531112 bytes in 0.264s)

running 16 tests
test deregister_without_register ... ok
test double_register_different_token ... ok
test is_send_and_sync ... ok
test poll_erroneous_registration ... ok
test drop_cancels_interest_and_shuts_down ... ok
test poll_registration ... ok
test poll_ok_after_cancelling_pending_ops ... ok
test registry_behind_arc ... ok
test poll_closes_fd ... ok
test reregister_interest_token_usage ... ok
test reregister_without_register ... ok
test registry_operations_are_thread_safe ... ok
test zero_duration_polls_events ... ok
test add_then_drop ... ok
test run_once_with_nothing ... ok
test register_during_poll ... ok

test result: ok. 16 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.21s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/registering-b36ebe61abe69842`
/Users/js/mio/target/i686-linux-android/debug/deps/registering-b36ebe61abe69842: 1 file pushed, 0 skipped. 37.4 MB/s (6748088 bytes in 0.172s)

running 5 tests
test tcp_register_multiple_event_loops ... ok
test udp_register_multiple_event_loops ... ok
test registering_after_deregistering ... ok
test register_deregister ... ok
test reregister_different_interest_without_poll ... ok

test result: ok. 5 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.41s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/regressions-f5e1d2ea19da9e1f`
/Users/js/mio/target/i686-linux-android/debug/deps/regressions-f5e1d2ea19da9e1f: 1 file pushed, 0 skipped. 34.9 MB/s (7067840 bytes in 0.193s)

running 3 tests
test issue_1403 ... ok
test issue_776 ... ok
test issue_1205 ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.51s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/tcp-7aa4fd1f7862d45c`
/Users/js/mio/target/i686-linux-android/debug/deps/tcp-7aa4fd1f7862d45c: 1 file pushed, 0 skipped. 40.9 MB/s (7995516 bytes in 0.187s)

running 18 tests
test bind_twice_bad ... ok
test connect_error ... ok
test accept ... ok
test connect_then_close ... ok
test connect ... ok
test is_send_and_sync ... ok
test local_addr_ready ... ok
test multiple_writes_immediate_success ... ok
test read ... ok
test peek ... ok
test listen_then_close ... ok
test connection_reset_by_peer ... ok
test write_error ... ok
test write_then_deregister ... ok
test write_then_drop ... ok
test write ... ok
test write_shutdown ... ok
test tcp_no_events_after_deregister ... ok

test result: ok. 18 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.41s

     Running `sh android-runner.sh /Users/js/mio/target/i686-linux-android/debug/deps/tcp_listener-22e243438b1cc2c4`
/Users/js/mio/target/i686-linux-android/debug/deps/tcp_listener-22e243438b1cc2c4: 1 file pushed, 0 skipped. 34.2 MB/s (7171500 bytes in 0.200s)

running 11 tests
test is_send_and_sync ... ok
test get_ttl_without_previous_set ... ok
test raw_fd ... ok
test set_get_ttl ... ok
test reregister ... ok
test no_events_after_deregister ... ok
test registering ... ok
test tcp_listener_two_streams ... ok
test tcp_listener_ipv6 ... test tcp_listener_ipv6 has been running for over 60 seconds
test tcp_listener ... test tcp_listener has been running for over 60 seconds
test tcp_listener_std ... test tcp_listener_std has been running for over 60 seconds
^C
```

**Comment by Thomasdezeeuw** [comment]
> Hi @Thomasdezeeuw! Sorry, I suppose I should have clarified: this issue happens on x86 Android. Attached is the test output, where the 3 tests hang indefinitely (`x86_64` tests all pass).
> 
> Are you saying that since it's _just_ Android x86 that's affected, it doesn't make sense to make these syscalls on the other platforms that are in the codepath (ios, macos, netbsd, solaris)? If so I can certainly gate it behind `#[cfg(and(target_arch="x86", target_os="android"))]`, I just figured that since both the docs I found for this syscall recommend always being explicit in setting the options, it would be the right thing to do. Thoughts?

We run tests on macOS (which shares a kernel with iOS) and it's known to work. NetBSD uses `accept4(2)`, which leaves Solaris. So limiting it to x86 Android would be best. 

> Android x86 test output (I manually killed the tests after a few minutes, they were clearly blocking):

I would expect the test to fail at this line: https://github.com/tokio-rs/mio/blob/527e843b0e11d585c0b20a8a3e4df966766f8bfa/tests/tcp_stream.rs#L72 Which calls https://github.com/tokio-rs/mio/blob/527e843b0e11d585c0b20a8a3e4df966766f8bfa/tests/util/mod.rs#L199-L207 If the socket was actually not non-blocking.

Can you conform that the non-blocking assertion is passed, because if that's the case it would be a Android kernel bug really.

**Comment by jshearer** [comment]
> Can you conform that the non-blocking assertion is passed
Will try that now

> because if that's the case it would be a Android kernel bug really.
Really? The linux `accept()` docs do say that the socket does _not_ inherit these flags, and isn't android a derivative of linux?

**Comment by Thomasdezeeuw** [comment]
> > Can you conform that the non-blocking assertion is passed
> 
> Will try that now
> 
> > because if that's the case it would be a Android kernel bug really.
> 
> Really? The linux `accept()` docs do say that the socket does _not_ inherit these flags, and isn't android a derivative of linux?

That's correct, the BSDs (FreeBSD, macOS, OpenBSD etc.) have this behaviour, but Linux doesn't. However the assertion in the test checks if the `TcpStream` is non-blocking, so if the kernel says it's non-blocking, while it's actually blocking that would be a kernel bug. 

> (forgive my ignorance here, I'm happy to be wrong and learn :) )

No worries, we're all learning and often wrong (that last thing might only apply to myself).

**Comment by jshearer** [comment]
... Fascinating. Indeed, the call to `assert_socket_non_blocking` passes just fine

**Comment by Thomasdezeeuw** [comment]
Well the test fails correctly as it's designed to catch this exact problem. Let's merge #1451 than as that is likely the quickest way to solve this. If you can open an issue with the Android (emulator?), I could make a small test case in C if needed.

**Comment by jshearer** [comment]
Okay cool. My fix in #1451 only addresses unix sockets, not TCP. I just made the equivalent change to `sys/unix/tcp.rs`, and now all the tests pass on x86 Android.

**Comment by Thomasdezeeuw** [comment]
> Okay cool. My fix in #1451 only addresses unix sockets, not TCP. I just made the equivalent change to `sys/unix/tcp.rs`, and now all the tests pass on x86 Android.

👍 

> Do you want me to gate this for x86-android only? Is it a... performance concern otherwise maybe (unneccesary syscall)?

Yes, I've reviewed #1451.
